### ISSUE TITLE
0.0.0.0 is functionally equivalent to 127.0.0.1/localhost, this project should use 0.0.0.1

### ISSUE THREAD

**Issue by pallaswept** [StevenBlack/hosts#2447]
In accordance with IANA and the appropriate RFCs, 0.0.0.0 aka 0.0.0.0/32 is "this host on this network" - in other words, it's locahost. You're already aware of this, although some of the issues raised with regards to this problem are so old I wouldn't blame you if it's slipped your memory.

> I still feel you make a valid point.  
_Originally posted by @StevenBlack in https://github.com/StevenBlack/hosts/issues/72#issuecomment-183584778_

> EDIT turns out 0.0.0.0/32 is 0.0.0.0, amiright?

According to the most recent documentation at https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml
0.0.0.0/32 is "this host on this network", and is distinguishable from 127.0.0.1 notably (in this context)  as 0.0.0.0 is not valid as a destination IP address, only as a source, and the RFCs state that it MUST be silently discarded if it does not meet that criteria. So, it should be enough, and suitable for use in as block list - but there is an exception, which is that 0.0.0.0 may be used as a destination IP for discovery of the local IP address. Which is why you have a bunch of cases stretching back about 7 years talking about how it works just like 127.0.0.1 and redirects to the local host just like they'd used 127.0.0.1. Here's the relevant portion of the RFC https://datatracker.ietf.org/doc/html/rfc1122#page-29, section 3.2.1.3 applies here

This describes 0.0.0.0/32
```
            (a)  { 0, 0 }

                 This host on this network.  MUST NOT be sent, except as
                 a source address as part of an initialization procedure
                 by which the host learns its own IP address.

                 See also Section 3.3.6 for a non-standard use of {0,0}.
```

This refers to 0.0.0.* aka 0.0.0.0/8
```
            (b)  { 0, <Host-number> }

                 Specified host on this network.  It MUST NOT be sent,
                 except as a source address as part of an initialization
                 procedure by which the host learns its full IP address.

```
Accordingly, this is still as valid as 0.0.0.0, but only if sent to another host interface on the same network, which - unless somebody has gone to the trouble of creating an interface on this same NIC, with a host IP address 0.0.0.* which could never be the destination of any traffic - will never happen.

You were already on the right track here:

>  I'm leaning towards following [RFC 3330](https://tools.ietf.org/html/rfc3330) and dish to an address in 0.0.0.0/32 other than 0.0.0.0.
_Originally posted by @StevenBlack in https://github.com/StevenBlack/hosts/issues/72#issuecomment-183586146_

> Presuming resolution times are the same, which I suspect will be true.

You presume correctly - the packets are 'silently discarded' as they have a destination IP address which is not allowed, other than that one exception of a configuration which would not function for any receipt of data.

You were very close here, but you gave up on this when you realised that 0.0.0.0/32 is 0.0.0.0. What you meant to do was to "dish to an address in 0.0.0.0/**_8_** (note: not /32) other than 0.0.0.0". As these other addresses (0.0.0.1-0.0.0.253) are not used as the initialisation procedure for discovering the loopback address (127.0.0.1), they are discarded as desired for a blocklist - with the added benefit over 0.0.0.0 which meets the exception criteria and not discarded but instead used to find and return the loopback address. There's supporting doco here: https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml

...

### In plain English: 
0.0.0.0 is valid as a destination address only when used as a mechanism to find out the loopback address - ie 0.0.0.0 exists only to find out what 127.0.0.1 is, so can be functionally identical to using 127.0.0.1
0.0.0.* is valid as only for a host to use as its own source address as a mechanism to allow another host to find out its own IP address - think DHCP -so it won't work like 127.0.0.1

[This poster found this answer independently](https://github.com/StevenBlack/hosts/issues/1974#issuecomment-1152136745) but you closed the case under the impression his was a system-configuration-specific fault - but as you can see from the documentation I've linked above, it's not, it's working correctly as per IANA and their RFCs. Not ALL systems MUST run this way, but it is a legit config with a legit problem, and the really bad news for this project is that **it's stock on linux.**, so for all intents and purposes, using 0.0.0.0 is the same as using 127.0.0.1 and will redirect blocklist entries to the localhost rather than blocking them. Furthermore reverse lookups to the loopback address can return hostnames from the blocklist (Yes, I discovered this issue when I had an "OMG why is my PC calling the locahost IP www.adsaboutgamblingonporn.com?!" moment of panic)

Take another look at his fix - use 0.0.0.1. It's not valid as a destination IP (so will serve the purpose of the block list here when outgoing packets with that destination IP are silently discarded) and is only valid as a source IP that is not implemented outside of certain rare enterprise networking equipment which use it as their source, and even still would not interfere with such devices as it could never transmit to them. 

### In plain English again: 
0.0.0.1 as good as 0.0.0.0 but better because unlike 0.0.0.0 it is never going to be the same as 127.0.0.1 and redirect all the  gambling hostnames to your local web server and make your desktop call itself pronhub.com

You were on the right track but bailed a bit before you got to the answer and mistook it as a system/user-specific issue, but it's not, it's standards implemented properly, and you already knew the resolution but just didn't notice (you tripped over on the wrong subnet and then blamed the end-user when you couldn't reproduce the bug locally, but there are several cases here to show that it's common).

I think you should use the workaround the above user and myself are using, and that you thought up way back then - use 0.0.0.1 instead of 0.0.0.0.
edit: or 0.0.0.42 or *.69 or *.187 whatever takes your fancy. 😆

**Comment by StevenBlack** [comment]
Thank you for this @pallaswept.

This isn't going to change. 

What you describe is a problem that evidently doesn't exist. If the semantics of `0.0.0.0` are bothersome, we have a `--ip` flag among our [command line options](https://github.com/StevenBlack/hosts#command-line-options) to generate the IP you desire, or require. 

For now, the default `0.0.0.0` is fine because nothing appears broken, as far as I know.

**Comment by StevenBlack** [comment]
@pallaswept I'd like to add this clarification.

This ship has sailed.

It might be a different if this was a repository with few followers, and few downstream consequences.

We have an installed base, here. Downstream there are countless services and applications that use what we publish for a surprisingly large number of ancillary uses. People rely on what this repository regularly produces.

A change like the one you suggest isn't something to take lightly, especially when it's pretty clear that nothing is broken, it fixes nothing, but downstream consequences are incalculable.  So it feels gratuitous, and I don't want to go there.

**Comment by pallaswept** [comment]
> nothing is broken

If this isn't broken:
> redirect all the gambling hostnames to your local web server and make your server call itself pronhub.com

What is? 

Because that's what 0.0.0.0 does on linux. I didn't imagine this problem up, I came across it when I did ping 127.0.0.1 and it came back with a porn website's name as my local host name. I thought I had some malware and investigating it led me to this.

> 0.0.0.0 is fine because nothing appears broken, as far as I know.

You know better because of all the cases you've had over the years where people have shown you.

> a problem that evidently doesn't exist

Get yourself an ubuntu live iso, boot it as a VM, and try it, if you want to see the evidence personally:
```

PING 0.0.0.0 (127.0.0.1) 56(84) bytes of data.
64 bytes from 127.0.0.1: icmp_seq=1 ttl=64 time=0.037 ms
64 bytes from 127.0.0.1: icmp_seq=2 ttl=64 time=0.039 ms
64 bytes from 127.0.0.1: icmp_seq=3 ttl=64 time=0.024 ms
64 bytes from 127.0.0.1: icmp_seq=4 ttl=64 time=0.025 ms
64 bytes from 127.0.0.1: icmp_seq=5 ttl=64 time=0.022 ms
64 bytes from 127.0.0.1: icmp_seq=6 ttl=64 time=0.021 ms
^C
--- 0.0.0.0 ping statistics ---
6 packets transmitted, 6 received, 0% packet loss, time 5060ms
rtt min/avg/max/mdev = 0.021/0.028/0.039/0.007 ms
```
```
linux@localhost:~> ping 0.0.0.1
PING 0.0.0.1 (0.0.0.1) 56(84) bytes of data.
^C
--- 0.0.0.1 ping statistics ---
8 packets transmitted, 0 received, 100% packet loss, time 7098ms
```

0.0.0.0 IS 127.0.0.1.

**Comment by pallaswept** [comment]
> It might be different if this was a repository with few followers, and few downstream consequences.
> We have an installed base, here. Downstream there are countless services and applications that use what we publish for a surprisingly large number of ancillary uses. People rely on what this repository regularly produces.
> A change like the one you suggest isn't something to take lightly, 

This part is true.

So stop taking it lightly.

**Comment by StevenBlack** [comment]
Hey Anon asshole @pallaswept let me make something perfectly clear to you.

Let me tell you where you stand:  One person out there who maybe wrote a bash script eight years ago that uses our files as input, trumps you.

When you've got a track record, when you create and maintain stuff for over a decade that serves millions of users, when you have an unimaginable installed base and you're loyal to them, then MAYBE you'll understand what's at play, and MAYBE then you'll have cred to talk to me like that.

**Comment by pallaswept** [comment]
You said I'm anon and have no track record; if I'm anon, then you can't know that. You're wrong again. You're calling me an asshole, which is ironic.
